Quickstart guides for Packet, AWS and Bare Metal (#60)
kinvolk/lokomotive: docs/quickstarts/packet.md

* [Step 3: Set up credentials from environment variables](#step-3-set-up-credentials-from-environment-variables)
* [Step 4: Define cluster configuration](#step-4-define-cluster-configuration)
* [Step 5: Create Lokomotive cluster](#step-5-create-lokomotive-cluster)
* [Verification](#verification)
* [Cleanup](#cleanup)
* [Troubleshooting](#troubleshooting)
* [Conclusion](#conclusion)
* [Next steps](#next-steps)

## Introduction

This quickstart guide walks through the steps needed to create a Lokomotive cluster on Packet with
Flatcar Container Linux using Route53 as the DNS provider.

By the end of this guide, you'll have a production-ready Kubernetes cluster running on Packet.

## Requirements

* Basic understanding of Kubernetes concepts.
* Packet account, Project ID and auth token (sometimes also referred to as [User Level API
  key](https://www.packet.com/developers/docs/API/getting-started/)).
* AWS account and IAM credentials (optional for Route53 DNS configuration).
* AWS Route53 DNS Zone (registered Domain Name or delegated subdomain).
* Terraform v0.12.x and [terraform-provider-ct](https://github.com/poseidon/terraform-provider-ct)
  installed locally.

> * Local BGP enabled.

Reviewer: Could we elaborate that, what it is?

Author: added a link for "how to enable local BGP in Packet"

guide.

### Step 2: Set up a working directory

It's better to start fresh in a new working directory, as the state of the cluster is stored in this
directory.

This also makes the cleanup task easier.

```console
mkdir -p lokomotive-infra/mycluster
cd lokomotive-infra/mycluster
```

### Step 3: Set up credentials from environment variables

#### Packet

* Log in to your Packet account and obtain the Project ID from the `Project Settings` tab.
* Obtain an API key from the `User Settings` menu.
* Set the environment variable `PACKET_AUTH_TOKEN` with the API key.

```console
export PACKET_AUTH_TOKEN=<PACKET_API_KEY>
```
#### AWS

Lokomotive requires AWS credentials for configuring Route53 DNS. To manually configure DNS entries
refer the DNS configuration settings for Packet(Add link).

```console
export AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE
export AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
```

### Step 4: Define cluster configuration

To create a Lokomotive cluster, we need to define a configuration.

A [production-ready configuration](../../examples/packet-production) is already provided for ease of

> use.  Copy the example configuration to the working directory and modify accordingly.

Reviewer: ```suggestion
use. Copy the example configuration to the working directory and modify accordingly.
```

+ ditto about modification

Node                                          Ready    Reason          Message

mycluster-controller-0                        True     KubeletReady    kubelet is posting ready status
mycluster-controller-1                        True     KubeletReady    kubelet is posting ready status
mycluster-controller-2                        True     KubeletReady    kubelet is posting ready status
mycluster-worker-pool-1-worker-0              True     KubeletReady    kubelet is posting ready status
mycluster-worker-pool-1-worker-1              True     KubeletReady    kubelet is posting ready status
mycluster-worker-pool-1-worker-2              True     KubeletReady    kubelet is posting ready status

Success - cluster is healthy and nodes are ready!
```

Use the generated `kubeconfig` file to access the Kubernetes cluster and list nodes.

```console
export KUBECONFIG=./lokomotive-assets/auth/kubeconfig
kubectl get nodes
```

## Cleanup

To destroy the Lokomotive cluster, execute the following command:

```console
lokoctl cluster destroy --confirm
```

You can safely delete the working directory created for this quickstart guide if you no longer
require it.

> ## Troubleshooting

Reviewer: I think this section should also include mention of insufficient availability of node types on Packet. That sometimes you may get an error, that there is no machines of type X available in region and then it's best to change the region or machine type used and refer to availability page of Packet (if one exists, I know it's available from the API).

We could also mention, that node types with spare disks should be used, otherwise installing prometheus operator component will fail.

### Stuck at copy controller secrets

If there is an execution error or no progress beyond the output provided below:

```console
...
module.packet-mycluster.null_resource.copy-controller-secrets: Still creating... (8m30s elapsed)
module.packet-mycluster.null_resource.copy-controller-secrets: Still creating... (8m40s elapsed)
...
```

The error is because the `ssh_pubkeys` provided in the configuration is missing in the `ssh-agent`

To rectify the error, you need to:

1. Execute [Cleanup](#cleanup) steps.
2. Follow the steps [to add the SSH key to the
   ssh-agent.](https://help.github.com/en/github/authenticating-to-github/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent#adding-your-ssh-key-to-the-ssh-agent)
3. Retry from Step 4:

### Packet provisioning failed

For failed machine provisioning on Packet end, execute the steps mentioned in the

> [Cleanup](#cleanup) section and retry [Step 5.](#step-5-create-lokomotive-cluster)

Reviewer: If provisioning fails, you don't need to destroy everything. You can just re-run `install` command and it will re-run Terraform.

In the event of failed Packet provisioning due to machines of variable `workers_type` not available.
You can check the Packet API [capacity endpoint](https://www.packet.com/developers/api/capacity/) to
get the current capacity and decide on changing the facility or the machine type.

### Permission issues

  * If the failure is due to insufficient permissions on Packet, check the permission on the Packet
    console.
  * This generally happens if user is using `Project Level API Key` and not `User Level API Key`.

### Failed installation of components that require disk storage

For components that require disk storage such as [Openebs storage
class](../configuration-guides/components/openebs-storage-class.md), [Prometheus
Operator](../configuration-guides/components/prometheus-operator.md) machine types with spare disks
should be used.

## Conclusion

After walking through this guide, you've learned how to set up a Lokomotive cluster on Packet.

## Next steps

You can now start deploying your workloads on the cluster.

For more information on installing supported Lokomotive components, you can visit the [component
configuration guides.](../configuration-guides/components)

> 

Reviewer 2: ```suggestion
```

# Lokomotive Packet installation guide

## Contents

* [Introduction](#introduction)
* [Requirements](#requirements)
* [Step 1: Install lokoctl](#step-1-install-lokoctl)
* [Step 2: Set up a temporary working directory](#step-2-set-up-a-temporary-working-directory)
* [Step 3: Set up credentials from environment variables](#step-3-set-up-credentials-from-environment-variables)
* [Step 4: Define cluster configuration](#step-4-define-cluster-configuration)
* [Step 5: Create Lokomotive cluster](#step-5-create-lokomotive-cluster)
* [Verification](#verification)
* [Cleanup](#cleanup)
* [Troubleshooting](#troubleshooting)
* [Conclusion](#conclusion)
* [Next steps](#next-steps)

## Introduction

This quickstart guide walks through the steps needed to create a Lokomotive cluster on Packet with
Flatcar Container Linux using Route53 as the DNS provider.

> By the end of this guide, you'll have a working Kubernetes cluster with 3 worker nodes and 3

Reviewer: Should we avoid mentioning specific settings here, which could be changed by the user?  
```suggestion
By the end of this guide, you'll have a  production-ready Kubernetes cluster running on Packet.
```

controller nodes.

## Requirements

* Basic understanding of Kubernetes concepts.
* Packet account, Project ID and auth token (sometimes also referred to as [User Level API
  key](https://www.packet.com/developers/docs/API/getting-started/)).
* AWS account and IAM credentials (optional for Route53 DNS configuration).
* AWS Route53 DNS Zone (registered Domain Name or delegated subdomain).
* Terraform v0.12.x and [terraform-provider-ct](https://github.com/poseidon/terraform-provider-ct)
  installed locally.
* Local BGP enabled.
* An SSH key pair for management access.
* `kubectl` installed locally to access the Kubernetes cluster.

## Steps

### Step 1: Install lokoctl

Lokoctl is a command-line interface for Lokomotive.

To install, find the [appropriate package](link-to-github-releases-page) for your system
and download it. Lokoctl is packaged as a `tar.gz` archive.

```console
wget <URL-TO-DOWNLOAD-LOKOCTL>
```
After downloading, untar the package. Lokoctl runs as a single binary named `lokoctl`

```console
tar xvzf <LOKOCTL_TAR_GZ_ARCHIVE>
```

> The final step is to make sure that the `lokoctl` binary is available on the `PATH`

Reviewer: This is technically not required, but makes it easier for further referenced commands. Can we maybe mention that somehow?

```console
export PATH=$PATH:/path/to/lokoctl/binary
```

> ### Step 2: Set up a temporary working directory

Reviewer: It's not temporary, as it contains cluster state.

It's better to start fresh in a temporary working directory, which makes the cleanup task easier.

```console
mkdir -p lokomotive-infra/mycluster
cd lokomotive-infra/mycluster
```

### Step 3: Set up credentials from environment variables

#### Packet

* Log in to your Packet account and obtain the Project ID from the `Project Settings` tab.
* Obtain an API key from the `User Settings` menu.
* Set the environment variable `PACKET_AUTH_TOKEN` with the API key.

```console
export PACKET_AUTH_TOKEN=<PACKET_API_KEY>
```
#### AWS

Lokomotive requires AWS credentials for configuring Route53 DNS. To manually configure DNS entries
refer the DNS configuration settings for Packet(Add link).

```console
export AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE
export AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
```

### Step 4: Define cluster configuration

To create a Lokomotive cluster, we need to define a configuration.

Create a file with the extension `.lokocfg` with the below contents.

```tf

> # mycluster.lokocfg

Reviewer: Please refer here to `examples` configuration and don't include the snippet here. It should mention `lokocfg.vars` example instead.

module.packet-mycluster.null_resource.bootkube-start: Creation complete after 4m25s [id=1122239320434737682]

Apply complete! Resources: 74 added, 0 changed, 0 destroyed.

Your configurations are stored in /home/imran/lokoctl-assets/mycluster

Now checking health and readiness of the cluster nodes ...

Node                                          Ready    Reason          Message

mycluster-controller-0                        True     KubeletReady    kubelet is posting ready status
mycluster-controller-1                        True     KubeletReady    kubelet is posting ready status
mycluster-controller-2                        True     KubeletReady    kubelet is posting ready status
mycluster-worker-pool-1-worker-0              True     KubeletReady    kubelet is posting ready status
mycluster-worker-pool-1-worker-1              True     KubeletReady    kubelet is posting ready status
mycluster-worker-pool-1-worker-2              True     KubeletReady    kubelet is posting ready status

Success - cluster is healthy and nodes are ready!
```

Use the generated `kubeconfig` file to access the Kubernetes cluster and list nodes.

```console

> export KUBECONFIG=$HOME/lokoctl-assets/mycluster/cluster-assets/auth/kubeconfig

Reviewer: This is inconsistent with previously mentioned working directory.
